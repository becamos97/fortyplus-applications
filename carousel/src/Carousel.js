
import React, { useState } from "react";
import Card from "./Card";
// import image data if needed
// import { images } from "./imageData";

function Carousel({ photos /* = images */ }) {
  const total = photos.length;
  const [currIdx, setCurrIdx] = useState(0);

  const goForward = () => setCurrIdx(i => Math.min(i + 1, total - 1));
  const goBackward = () => setCurrIdx(i => Math.max(i - 1, 0));

  const currPhoto = photos[currIdx];

   return (
    <div className="Carousel">
      {/* Only show left arrow if not on first */}
      {currIdx > 0 && (
        <i
          className="fas fa-chevron-circle-left"
          data-testid="left-arrow"
          onClick={goBackward}
          aria-label="previous"
        />
      )}

      <Card
        caption={currPhoto.caption}
        src={currPhoto.src}
        currNum={currIdx + 1}
        totalNum={total}
      />

      {/* Only show right arrow if not on last */}
      {currIdx < total - 1 && (
        <i
          className="fas fa-chevron-circle-right"
          data-testid="right-arrow"
          onClick={goForward}
          aria-label="next"
        />
      )}
    </div>
  );
}

export default Carousel;