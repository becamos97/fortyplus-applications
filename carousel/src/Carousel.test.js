import React from "react";
import { render, fireEvent } from "@testing-library/react";
import Carousel from "./Carousel";
import TEST_IMAGES from "./_testCommon.js";

// helper: find right arrow by data-testid or common classes
function getRightArrow(container) {
  return (
    container.querySelector('[data-testid="right-arrow"]') ||
    container.querySelector(".bi-arrow-right-circle") ||
    container.querySelector(".bi-arrow-right-circle-fill") ||
    container.querySelector(".right-arrow") ||
    null
  );
}

// helper: find left arrow by data-testid or common classes
function getLeftArrow(container) {
  return (
    container.querySelector('[data-testid="left-arrow"]') ||
    container.querySelector(".bi-arrow-left-circle") ||
    container.querySelector(".bi-arrow-left-circle-fill") ||
    container.querySelector(".left-arrow") ||
    null
  );
}

describe("Carousel", () => {
  // Smoke
  it("renders without crashing", () => {
    render(<Carousel photos={TEST_IMAGES} title="images for testing" />);
  });

  // Snapshot
  it("matches snapshot", () => {
    const { asFragment } = render(
      <Carousel photos={TEST_IMAGES} title="images for testing" />
    );
    expect(asFragment()).toMatchSnapshot();
  });

  // Demo test (right arrow moves forward)
  it("works when you click on the right arrow", function () {
    const { container } = render(
      <Carousel photos={TEST_IMAGES} title="images for testing" />
    );
    expect(container.querySelector('img[alt="testing image 1"]')).toBeInTheDocument();
    expect(container.querySelector('img[alt="testing image 2"]')).not.toBeInTheDocument();

    const rightArrow = getRightArrow(container);
    expect(rightArrow).toBeTruthy(); // ensure we found it
    fireEvent.click(rightArrow);

    expect(container.querySelector('img[alt="testing image 1"]')).not.toBeInTheDocument();
    expect(container.querySelector('img[alt="testing image 2"]')).toBeInTheDocument();
  });

  // Part 3: left arrow should go BACK
  it("moves backward when left arrow is clicked (from second image)", () => {
    const { container } = render(
      <Carousel photos={TEST_IMAGES} title="images for testing" />
    );

    // go to second image
    const rightArrow = getRightArrow(container);
    expect(rightArrow).toBeTruthy();
    fireEvent.click(rightArrow);

    expect(container.querySelector('img[alt="testing image 2"]')).toBeInTheDocument();

    const leftArrow = getLeftArrow(container);
    expect(leftArrow).toBeTruthy();
    fireEvent.click(leftArrow);

    expect(container.querySelector('img[alt="testing image 1"]')).toBeInTheDocument();
    expect(container.querySelector('img[alt="testing image 2"]')).not.toBeInTheDocument();
  });

  // Part 4: hide left on first, hide right on last
  it("hides left arrow on first image and right arrow on last image", () => {
    const { container } = render(
      <Carousel photos={TEST_IMAGES} title="images for testing" />
    );

    // On first image: left arrow should be ABSENT
    expect(getLeftArrow(container)).toBeNull();

    // Click right until right arrow disappears (now at last image)
    let right = getRightArrow(container);
    // right should exist initially; if not, fail fast:
    expect(right).toBeTruthy();
    while (right) {
      fireEvent.click(right);
      right = getRightArrow(container);
    }

    // Now on last image: right arrow should be ABSENT
    expect(getRightArrow(container)).toBeNull();

  });
});