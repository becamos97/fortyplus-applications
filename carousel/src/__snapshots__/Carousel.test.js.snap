// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Carousel matches snapshot 1`] = `
<DocumentFragment>
  <div
    class="Carousel"
  >
    <div
      class="Card"
    >
      <h4
        class="Card-title"
      >
        testing image 1
      </h4>
      <img
        alt="testing image 1"
        class="Card-image"
        src="test1.com"
      />
      <small
        class="Card-small"
      >
        Image 1 of 3.
      </small>
    </div>
    <i
      aria-label="next"
      class="fas fa-chevron-circle-right"
      data-testid="right-arrow"
    />
  </div>
</DocumentFragment>
`;
